internal/gcs/container.go

```diff
@@ -0,0 +1,171 @@
+package gcs
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/logfields"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	hrComputeSystemDoesNotExist = 0xc037010e
+)
+
+func (gc *GuestConnection) requestNotify(cid string, ch chan struct{}) error {
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	if gc.notifyChs == nil {
+		return errors.New("guest connection closed")
+	}
+	if _, ok := gc.notifyChs[cid]; ok {
+		return fmt.Errorf("container %s already exists", cid)
+	}
+	gc.notifyChs[cid] = ch
+	return nil
+}
+
+func (gc *GuestConnection) notify(ntf *containerNotification) error {
+	cid := ntf.ContainerID
+	gc.mu.Lock()
+	ch := gc.notifyChs[cid]
+	delete(gc.notifyChs, cid)
+	gc.mu.Unlock()
+	if ch == nil {
+		return fmt.Errorf("container %s not found", cid)
+	}
+	logrus.WithField(logfields.ContainerID, cid).Info("container terminated in guest")
+	close(ch)
+	return nil
+}
+
+func (gc *GuestConnection) clearNotifies() {
+	gc.mu.Lock()
+	chs := gc.notifyChs
+	gc.notifyChs = nil
+	gc.mu.Unlock()
+	for _, ch := range chs {
+		close(ch)
+	}
+}
+
+type Container struct {
+	gc        *GuestConnection
+	id        string
+	notifyCh  chan struct{}
+	closeCh   chan struct{}
+	closeOnce sync.Once
+}
+
+func (gc *GuestConnection) CreateContainer(ctx context.Context, cid string, cfg interface{}) (*Container, error) {
```

Comment:
Can we put the `gc *GuestConnection) *` methods in `guestconnection.go`?

Reply:
If you want... to me, it's clearer to have container.go be the only place that constructs a `Container` so that the type and the initialization of it is in the same place.

Reply:
My intellisense pattern seems different than yours. The first thing I do for a type is "go to file" named that type. Often that gets me very close to the method's of that type. What do you do? Search for CreateContainer in the package?

Reply:
Yes. I guess I rely more on search and Ctrl-click with vscode than files. But once I'm in a file, I want everything related to be there; I don't want to jump around between files.

Reply:
Spoke offline. Leaving for now